### PR TITLE
Type definitions

### PR DESCRIPTION
**Pull request by uetchy**
# Summary

Added.ts` files fo each package.

- [x] remark-math
- [x] rehype-katex
- [x] rehype-mathjax
- [x] remark-html-katex

## Discussion

- Should `options` have index signature `[index: string]: unknown` as `unified.Settings` did?
  - pros: can pass additional options like `{position: boolean}`
    - another way to deal with it is to explicitly add `{position: boolean}` to the definition files.
  - cons: it will disable key missing errors that help users noticing a typo.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -15,6 +15,7 @@
   ],
   "dependencies": {},
   "devDependencies": {
+    "eslint-plugin-prettier": "^3.1.4",
```
Comment (ChristianMurphy) [inline package.json:18]: is this dependency needed?

File: packages/rehype-mathjax/types/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  mathmlSpacing: boolean
+  skipAttributes: {[index: string]: boolean}
+  exFactor: number
+  displayAlign: 'left' | 'center' | 'right'
+  displayIndent: string
+  fontCache: 'local' | 'global'
+  localID: string | null
+  internalSpeechTitles: true
+  titleID: number
+}
```
Comment (ChristianMurphy) [inline packages/rehype-mathjax/types/index.d.ts:19]: MathJax has typings https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/mathjax/index.d.ts which may be able to be referenced.

Also this package has three entrypoints https://github.com/remarkjs/remark-math/tree/82aeb9c3231070c15fc8ea6fa429ffd8677e54b3/packages/rehype-mathjax#rehypeuserehypemathjax-options

File: packages/remark-math/package.json [hunk 3]
```diff
@@ -33,7 +33,10 @@
     "util.js"
   ],
   "main": "index.js",
+  "types": "types/index.d.ts",
   "dependencies": {},
-  "devDependencies": {},
+  "peerDependencies": {
+    "unified": "^9.0.0"
```
Comment (ChristianMurphy) [inline packages/remark-math/package.json:39]: remark/unified generally doesn't use peerDependencies, they can cause trouble.
Generally `dependencies` or `devDependencies` are used.

File: packages/rehype-katex/package.json [hunk 4]
```diff
@@ -33,6 +33,7 @@
     "index.js"
```
Comment (ChristianMurphy) [inline packages/rehype-katex/package.json:33]: the types also need to be included in the package

```suggestion
    "index.js",
    "types/index.d.ts"
```

File: package.json (hunk 1 above)
Comment (uetchy) [reply package.json:18]: Without this line, `npm test` will fail.

File: packages/remark-math/package.json (hunk 3 above)
Comment (uetchy) [reply packages/remark-math/package.json:39]: ok. I'll move it to dependencies.

File: packages/rehype-mathjax/types/index.d.ts (hunk 2 above)
Comment (uetchy) [reply packages/rehype-mathjax/types/index.d.ts:19]: Looking around `@types/mathjax` but couldn't find SVGOptions there. So we have to define it after all.

Comment (uetchy) [reply packages/rehype-mathjax/types/index.d.ts]: @ChristianMurphy Added types for chtml and browser in https://github.com/remarkjs/remark-math/pull/48/commits/7ee8c0434b35b15f10db5b8f5c8f6ad1115c3202

**Comment by uetchy** [comment]
@ChristianMurphy I think all issues you mentioned are resolved now.

File: packages/rehype-mathjax/chtml.d.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+  minScale?: number
+  matchFontHeight?: boolean
+  mtextInheritFont?: boolean
+  merrorInheritFont?: boolean
+  mathmlSpacing?: boolean
+  skipAttributes?: {[index: string]: boolean}
+  exFactor?: number
+  displayAlign?: 'left' | 'center' | 'right'
+  displayIndent?: string
+  fontURL: string
```
Comment (ChristianMurphy) [inline packages/rehype-mathjax/chtml.d.ts:15]: is `fontURL` required? In the docs, it seems to indicate there is a default https://docs.mathjax.org/en/latest/options/output/chtml.html#the-configuration-block
Which implies it is optional, is that accurate?

File: packages/rehype-mathjax/browser.d.ts [hunk 6]
```diff
@@ -0,0 +1,12 @@
+import {Plugin} from 'unified'
+
+// http://docs.mathjax.org/en/latest/options/input/tex.html#the-configuration-block
+type MathNotation = [string, string]
+interface BrowserOptions {
+  displayMath: MathNotation | MathNotation[]
+  inlineMath: MathNotation | MathNotation[]
+}
+
+declare const renderBrowser: Plugin<[Partial<BrowserOptions>?]>
```
Comment (ChristianMurphy) [inline packages/rehype-mathjax/browser.d.ts:10]: There are some different approach to options across the unified/remark community.
We are normalizing toward using `?` for optional parameters in most cases, and using `Partial` only for external types that need to be made optional.
Since this (currently) an internal interface, `?` would be preferred.

File: packages/rehype-mathjax/index.d.ts [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+  mathmlSpacing: boolean
+  skipAttributes: {[index: string]: boolean}
+  exFactor: number
+  displayAlign: 'left' | 'center' | 'right'
+  displayIndent: string
+  fontCache: 'local' | 'global'
+  localID: string | null
+  internalSpeechTitles: true
+  titleID: number
+}
```
Comment (ChristianMurphy) [inline packages/rehype-mathjax/index.d.ts:19]: I'm cautious about maintaining a copy of the MathJax typings.
I'd lean towards getting this added upstream, them incorporating the update here.

thoughts @wooorm ? :thought_balloon:

File: package.json (hunk 1 above)
Comment (ChristianMurphy) [reply package.json:19]: That is surprising, `xo` is the `eslint` wrapper and it is supposed to automatically bundle all its dependencies.
What error are you seeing when this dependency isn't explicitly added?

File: packages/rehype-mathjax/index.d.ts (hunk 7 above)
Comment (wooorm) [reply packages/rehype-mathjax/index.d.ts:19]: It would be great if we can import them, indeed!

File: packages/rehype-mathjax/chtml.d.ts (hunk 5 above)
Comment (wooorm) [reply packages/rehype-mathjax/chtml.d.ts:15]: The default `fontURL` does not make sense in our case, as we’re processing on the server, so for this project it is required. See also: <https://github.com/remarkjs/remark-math/pull/45/commits/0c4505c44635b6cd3c7fc49b408a985ac68684f4#r422095782>

File: packages/rehype-mathjax/index.d.ts (hunk 7 above)
Comment (uetchy) [reply packages/rehype-mathjax/index.d.ts:19]: @ChristianMurphy @wooorm Yes it should definitely be bundled in mathjax-full package. I'll open PR on MathJax repo to include these interfaces on their release.

File: package.json (hunk 1 above)
Comment (wooorm) [reply package.json:19]: @uetchy could you try 1) removing this line, 2) removing (nested) `node_modules`, 3) `npm install`ing again? And see if it works now?

**Comment by ChristianMurphy** [comment]
> Should options have index signature `[index: string]: unknown` as `unified.Settings` did?

Generally that is used when settings are generic.
In unified the index signature is a fallback if a package doesn't define it's own settings.
In this case, concrete settings would be preferred, listing out each attribute an its type.

File: packages/rehype-mathjax/browser.d.ts (hunk 6 above)
Comment (ChristianMurphy) [reply packages/rehype-mathjax/browser.d.ts:10]: @uetchy friendly ping

File: packages/rehype-mathjax/index.d.ts (hunk 7 above)
Comment (uetchy) [reply packages/rehype-mathjax/index.d.ts:19]: @ChristianMurphy I initially thought I would restart working on this PR after [this issue](https://github.com/mathjax/MathJax-src/issues/524) resolved, but there's no response for a long time. Is it ok to temporary bundle those interfaces into remark-math until it is included in MathJax?

Comment (wooorm) [reply packages/rehype-mathjax/index.d.ts:19]: Yeah, I think so!

File: package.json (hunk 1 above)
Comment (uetchy) [reply package.json:19]: I tried:

1. `rm -rf **/node_modules`
1. `yarn`
1. `node_modules/.bin/xo --fix`

and got this:

```
Error: Failed to load plugin 'prettier' declared in 'CLIOptions': Cannot find module 'eslint-plugin-prettier'
Require stack:
- /Users/uetchy/Repos/src/github.com/remarkjs/remark-math/__placeholder__.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:1085:15)
    at Function.resolve (internal/modules/cjs/helpers.js:81:19)
    at Object.resolve (/Users/uetchy/Repos/src/github.com/remarkjs/remark-math/node_modules/eslint/lib/shared/relative-module-resolver.js:28:50)
    at ConfigArrayFactory._loadPlugin (/Users/uetchy/Repos/src/github.com/remarkjs/remark-math/node_modules/eslint/lib/cli-engine/config-array-factory.js:992:39)
    at /Users/uetchy/Repos/src/github.com/remarkjs/remark-math/node_modules/eslint/lib/cli-engine/config-array-factory.js:883:33
    at Array.reduce (<anonymous>)
    at ConfigArrayFactory._loadPlugins (/Users/uetchy/Repos/src/github.com/remarkjs/remark-math/node_modules/eslint/lib/cli-engine/config-array-factory.js:879:22)
    at ConfigArrayFactory._normalizeObjectConfigDataBody (/Users/uetchy/Repos/src/github.com/remarkjs/remark-math/node_modules/eslint/lib/cli-engine/config-array-factory.js:707:44)
    at _normalizeObjectConfigDataBody.next (<anonymous>)
    at ConfigArrayFactory._normalizeObjectConfigData (/Users/uetchy/Repos/src/github.com/remarkjs/remark-math/node_modules/eslint/lib/cli-engine/config-array-factory.js:647:20)
```

and realized that I only get this error when I use `yarn`. I'll use `npm` for this PR as it has no issue around `xo`.

Comment (uetchy) [reply package.json]: fixed in https://github.com/remarkjs/remark-math/pull/48/commits/237fbe8599d6d420c54d7ab6d0d7a84e77b22c45

File: packages/rehype-mathjax/browser.d.ts (hunk 6 above)
Comment (uetchy) [reply packages/rehype-mathjax/browser.d.ts]: @ChristianMurphy Fixed now! https://github.com/remarkjs/remark-math/pull/48/commits/aef142dc8c9d275ec7296212b3653dd108de1986

**Comment by uetchy** [comment]
> Would it be possible to add some simple type tests for dtslint?
> [omit]
> for each typing

@ChristianMurphy Added type test for each package!

File: package.json [hunk 8]
```diff
@@ -36,7 +38,12 @@
     "format": "remark . -qfo && prettier . --write && xo --fix",
     "test-api": "tape \"packages/*/test.js\" \"packages/*/test/**/*.js\"",
     "test-coverage": "nyc --reporter lcov tape \"packages/*/test.js\" \"packages/*/test/**/*.js\"",
-    "test": "npm run format && npm run test-coverage"
+    "test:types": "run-p test:types:rehype-katex test:types:rehype-mathjax test:types:remark-html-katex test:types:remark-math",
+    "test:types:rehype-katex": "dtslint packages/rehype-katex/types",
+    "test:types:rehype-mathjax": "dtslint packages/rehype-mathjax",
+    "test:types:remark-html-katex": "dtslint packages/remark-html-katex/types",
+    "test:types:remark-math": "dtslint packages/remark-math/types",
+    "test": "run-s format test:types test-coverage"
```
Comment (wooorm) [inline package.json:46]: This seems a bit verbose to me. Would `dtslint packages/*/types` or so work too? If so, I’d prefer it instead of run-all and separate scripts.

File: packages/rehype-mathjax/index.d.ts [hunk 9]
```diff
@@ -0,0 +1,26 @@
+// Minimum TypeScript Version: 3.2
+import {Plugin} from 'unified'
+
+// Should be ported back to MathJax repo
+// http://docs.mathjax.org/en/latest/options/output/svg.html#the-configuration-block
+interface MathJaxSVGOptions {
```
Comment (wooorm) [inline packages/rehype-mathjax/index.d.ts:6]: For our own typings, could you do camelcase like `MathJaxSvgOptions` instead of ~`MathJaxSVGOptions`~?

File: packages/rehype-mathjax/package.json [hunk 10]
```diff
@@ -34,19 +34,22 @@
...
   ],
   "main": "index.js",
+  "types": "index.d.ts",
   "dependencies": {
+    "@types/mathjax": "^0.0.36",
     "hast-util-from-dom": "^2.0.0",
     "hast-util-to-text": "^2.0.0",
     "jsdom": "^16.0.0",
     "mathjax-full": "^3.0.0",
+    "unified": "^9.0.0",
```
Comment (wooorm) [inline packages/rehype-mathjax/package.json:51]: I don’t think we should put `unified` in dependencies, I suggest to put something in the docs that folks install it themselves.

File: packages/remark-html-katex/types/index.d.ts [hunk 11]
```diff
@@ -0,0 +1,8 @@
+import {Plugin} from 'unified'
+import {KatexOptions} from 'katex'
+
+type HTMLKatexOptions = KatexOptions
```
Comment (wooorm) [inline packages/remark-html-katex/types/index.d.ts:4]: ~`HTMLKatexOptions`~ -> `HtmlKatexOptions`

File: packages/remark-math/package.json (hunk 3 above)
Comment (wooorm) [reply packages/remark-math/package.json:39]: Sorry, I want to reopen this!

While I think including a package for types is fine (e.g., types/hast, vfile). I don’t think we should do it for unified. I think it should be removed from dependencies, and a note added in the readme that they should be installed (although: plugins don’t make sense w/o unified, so they probably always are installed).

File: package.json (hunk 8 above)
Comment (uetchy) [reply package.json:46]: The problem here is that only `rehype-mathjax` has multiple type definition files that must be placed in the same directory as the original file, and that `dtslint packages/*/types` will fail because there is no `types` dir in `rehype-mathjax`.
Possible solutions are:
1. Move `test-types` to each package and run `lerna run test-types` in the root.

**Comment by pd4d10** [comment]
Any news?
